external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisRecordProcessor.scala

```diff
@@ -56,6 +56,31 @@ private[kinesis] class KinesisRecordProcessor[T](receiver: KinesisReceiver[T], w
     logInfo(s"Initialized workerId $workerId with shardId $shardId")
   }
 
+  private def addRecords(batch: List[Record], checkpointer: IRecordProcessorCheckpointer): Unit = {
+    receiver.addRecords(shardId, batch)
+    logDebug(s"Stored: Worker $workerId stored ${batch.size} records for shardId $shardId")
+    receiver.setCheckpointer(shardId, checkpointer)
+  }
+
+  /**
+   * Limit the number of processed records from Kinesis stream. This is because the KCL cannot
+   * control the number of aggregated records to be fetched even if we set `MaxRecords`
+   * in `KinesisClientLibConfiguration`. For example, if we set 10 to the number of max records
+   * in a worker and a producer aggregates two records into one message, the worker possibly
+   * 20 records every callback function called.
+   */
+  private def processRecordsWithLimit(
+      batch: List[Record], checkpointer: IRecordProcessorCheckpointer): Unit = {
+    val maxRecords = receiver.getCurrentLimit
+    if (batch.size() <= maxRecords) {
+      addRecords(batch, checkpointer)
```

Comment:
I think the for loop even takes care of this case, but no big deal either way. It seems like a reasonable change.